onnxruntime/core/providers/cpu/ml/treeregressor.cc

```diff
@@ -79,149 +60,216 @@ TreeEnsembleRegressor<T>::TreeEnsembleRegressor(const OpKernelInfo& info)
   ORT_ENFORCE(nodes_id_size == nodes_truenodeids_.size());
   ORT_ENFORCE(nodes_id_size == nodes_falsenodeids_.size());
   ORT_ENFORCE((nodes_id_size == nodes_hitrates_.size()) || (nodes_hitrates_.empty()));
+  ORT_ENFORCE(base_values_.empty() || base_values_.size() == static_cast<size_t>(n_targets_));
 
   max_tree_depth_ = 1000;
-  offset_ = four_billion_;
-  using LeafNodeData = std::tuple<int64_t, int64_t, int64_t, float>;
-  //leafnode data, these are the votes that leaves do
-  for (size_t i = 0; i < target_nodeids_.size(); i++) {
-    leafnode_data_.push_back(std::make_tuple(target_treeids_[i], target_nodeids_[i], target_ids_[i], target_weights_[i]));
-  }
-  std::sort(begin(leafnode_data_), end(leafnode_data_), [](LeafNodeData const& t1, LeafNodeData const& t2) {
-    if (std::get<0>(t1) != std::get<0>(t2))
-      return std::get<0>(t1) < std::get<0>(t2);
-
-    return std::get<1>(t1) < std::get<1>(t2);
-  });
-  //make an index so we can find the leafnode data quickly when evaluating
-  int64_t field0 = -1;
-  int64_t field1 = -1;
-  for (size_t i = 0; i < leafnode_data_.size(); i++) {
-    int64_t id0 = std::get<0>(leafnode_data_[i]);
-    int64_t id1 = std::get<1>(leafnode_data_[i]);
-    if (id0 != field0 || id1 != field1) {
-      int64_t id = id0 * four_billion_ + id1;
-      auto p3 = std::make_pair(id, i);  // position is i
-      leafdata_map_.insert(p3);
-      field0 = id;
-      field1 = static_cast<int64_t>(i);
-    }
-  }
-  //treenode ids, some are roots, and roots have no parents
-  std::unordered_map<int64_t, size_t> parents;  //holds count of all who point to you
-  std::unordered_map<int64_t, size_t> indices;
-  //add all the nodes to a map, and the ones that have parents are not roots
-  std::unordered_map<int64_t, size_t>::iterator it;
-  size_t start_counter = 0L;
-  for (size_t i = 0; i < nodes_treeids_.size(); i++) {
-    //make an index to look up later
-    int64_t id = nodes_treeids_[i] * four_billion_ + nodes_nodeids_[i];
-    auto p3 = std::make_pair(id, i);  // i is the position
-    indices.insert(p3);
-    it = parents.find(id);
-    if (it == parents.end()) {
-      //start counter at 0
-      auto p1 = std::make_pair(id, start_counter);
-      parents.insert(p1);
-    }
+
+  // Filling nodes_.
+  nbnodes_ = nodes_treeids_.size();
+  nodes_ = new TreeNodeElement[(int)nbnodes_];
+  roots_.clear();
+  std::map<TreeNodeElementId, TreeNodeElement*> idi;
+  size_t i;
+
+  for (i = 0; i < nodes_treeids_.size(); ++i) {
+    TreeNodeElement* node = nodes_ + i;
```

Comment:
nit: setup end condition once instead of calling size() on every iteration. 
e.g.
 for (size_t i = 0, end = nodes_treeids_.size(); i < end; ++i)